drivers/gpio/gpio_aw9523b.c

```diff
@@ -0,0 +1,474 @@
+	}
+
+	/* Open-drain support is per port, not per pin.
+	 * So can't really support the API as-is.
+	 */
+	if (port == 0 && !config->port0_push_pull) {
+		if (!((flags & GPIO_SINGLE_ENDED) && (flags & GPIO_LINE_OPEN_DRAIN))) {
+			return -ENOTSUP;
+		}
+	} else {
+		if (flags & GPIO_SINGLE_ENDED) {
+			return -ENOTSUP;
+		}
+	}
+
+	if ((flags & GPIO_INPUT) && ((flags & GPIO_PULL_UP) || (flags & GPIO_PULL_DOWN))) {
+		return -ENOTSUP;
+	}
+
+	k_sem_take(aw9523b_get_lock(config->mfd_dev), K_FOREVER);
+
+	err = i2c_reg_update_byte_dt(&config->i2c, AW9523B_REG_CONFIG0 + port, mask, input_en);
+	if (err) {
+		LOG_ERR("%s: Failed to set pin%d direction (%d)", dev->name, pin, err);
+		goto on_error;
+	}
+
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
+	if (config->int_gpio.port) {
+		if (input_en) {
+			struct gpio_aw9523b_data *const data = dev->data;
+			uint8_t buf[2];
+
+			/* Read initial pin state */
+			err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 2);
+			if (err) {
+				LOG_ERR("%s: Read initial pin state failed (%d)", dev->name, err);
+				goto on_error;
+			}
+
+			WRITE_BIT(data->prev_value, pin, !!((buf[1] << 8 | buf[0]) & BIT(pin)));
+		} else {
+			struct gpio_aw9523b_data *const data = dev->data;
+
+			WRITE_BIT(data->falling_event_pins, pin, 0);
+			WRITE_BIT(data->rising_event_pins, pin, 0);
+		}
+	}
+#endif
+
+	err = i2c_reg_update_byte_dt(&config->i2c, AW9523B_REG_OUTPUT0 + port, mask, out_high);
+	if (err) {
+		LOG_ERR("%s: Failed to set initial pin state (%d)", dev->name, err);
+		return err;
+	}
+
+on_error:
+	k_sem_give(aw9523b_get_lock(config->mfd_dev));
+
+	return err;
+}
+
+static int gpio_aw9523b_port_read_write_toggle(const struct device *dev, gpio_port_pins_t mask,
+					       gpio_port_value_t *value,
+					       enum read_write_toggle_t mode)
+{
+	const struct gpio_aw9523b_config *const config = dev->config;
+	uint8_t buf[2];
+	int err;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(aw9523b_get_lock(config->mfd_dev), K_FOREVER);
+
+	err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 1);
+	if (!err) {
+		err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT1, buf + 1, 1);
+	}
+
+	if (err) {
+		LOG_ERR("%s: Failed to read port status (%d)", dev->name, err);
+	} else if (mode != READ) {
+		gpio_port_value_t old_value = (buf[1] << 8 | buf[0]);
+		gpio_port_value_t new_value;
+
+		if (mode == WRITE) {
+			new_value = (old_value & ~mask) | (*value & mask);
+		} else {
+			new_value = (old_value & ~mask) | (~old_value & mask);
+		}
+
+		if (new_value != old_value) {
+			buf[0] = (new_value >> 0) & 0xFF;
+			buf[1] = (new_value >> 8) & 0xFF;
+			err = i2c_burst_write_dt(&config->i2c, AW9523B_REG_OUTPUT0, buf, 2);
+			if (err) {
+				LOG_ERR("%s: Failed to set port (%d)", dev->name, err);
+			}
+		}
+	}
+
+	k_sem_give(aw9523b_get_lock(config->mfd_dev));
+
+	if (err == 0 && mode == READ) {
+		*value = (buf[1] << 8 | buf[0]);
```

Comment:
`sys_get_le16`, but also yeah it's hard to track what code path executes this, if you could refactor it to make all the breanches and exit paths obvious it'd be great

Reply:
I've reorganized this entire function. I think this is getting better.